### PR TITLE
[NNC] Support more IR in C++ codegen

### PR DESCRIPTION
**Pull request by ghost**
This PR introduces `Vector` and `Tensor` related classes and functions in `cpp_vector.h` and `cpp_tensor.h`, and implements C++ codegen for more IR nodes based on these abstractions. 

Vector operations are currently implemented as naive for loops, in the future, they should be implemented using SIMD intrinsics such as AVX.

### CODE REVIEW

**Review by ZolotukhinM — APPROVED** [review]
This looks good to me overall! One question that I have: would we be able to somehow test the generated C++ code? I'd imagine that it's easy to, say, make a typo there now, and we won't catch it with current tests.

Also, please make sure that all tests pass before landing - the failure in `test_jit_fuser_te.py` is probably real.

File: torch/csrc/jit/tensorexpr/ir_printer.cpp
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_printer.cpp:192]: Looks like this change causes miscompilation on cuda:
```
Feb 23 10:57:36 ======================================================================
Feb 23 10:57:36 ERROR [2.194s]: test_clamp (__main__.TestTEFuser)
Feb 23 10:57:36 ----------------------------------------------------------------------
Feb 23 10:57:36 Traceback (most recent call last):
Feb 23 10:57:36   File "test_jit_fuser_te.py", line 434, in test_clamp
Feb 23 10:57:36     s = self.checkScript(f, (inp1, inp2), profiling=ProfilingMode.PROFILING)
Feb 23 10:57:36   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/jit_utils.py", line 457, in checkScript
Feb 23 10:57:36     frames_up=2)
Feb 23 10:57:36   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/jit_utils.py", line 482, in checkScript
Feb 23 10:57:36     opt_script_outputs = scripted_fn(*recording_inputs)
Feb 23 10:57:36   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 129, in prof_func_call
Feb 23 10:57:36     return prof_callable(func_call, *args, **kwargs)
Feb 23 10:57:36   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 126, in prof_callable
Feb 23 10:57:36     return callable(*args, **kwargs)
Feb 23 10:57:36 RuntimeError: default_program(24): error: identifier "INFINITY" is undefined
Feb 23 10:57:36 
Feb 23 10:57:36 1 error detected in the compilation of "default_program".
```

Comment (cheng-chang) [reply torch/csrc/jit/tensorexpr/ir_printer.cpp:192]: I'll change it back to POS_INFINITY and NEG_INFINITY, didn't realize it's used for CUDA. In C++, we only have INFINITY and -INFINITY though.

**Comment by cheng-chang** [comment]
> This looks good to me overall! One question that I have: would we be able to somehow test the generated C++ code? I'd imagine that it's easy to, say, make a typo there now, and we won't catch it with current tests.
> 
> Also, please make sure that all tests pass before landing - the failure in `test_jit_fuser_te.py` is probably real.

I was also thinking about how to test the generated code, but didn't come up with a satisfying solution other than implement the codegen to compile and run the generated c++.

**Comment by cheng-chang** [comment]
I'll add support for Intrinsics and ExternalCall in this PR too.

**Comment by ZolotukhinM** [comment]
> I was also thinking about how to test the generated code, but didn't come up with a satisfying solution other than implement the codegen to compile and run the generated c++.

Yeah, I could not think of an elegant solution to this too. Maybe we can expose this functionality to python and write an end-to-end test there: e.g. convert a TE stmt to C++ and call a system c++ compiler on to see that it compiles - what do you think? (No need to do that in this PR)

**Comment by cheng-chang** [comment]
> > I was also thinking about how to test the generated code, but didn't come up with a satisfying solution other than implement the codegen to compile and run the generated c++.
> 
> Yeah, I could not think of an elegant solution to this too. Maybe we can expose this functionality to python and write an end-to-end test there: e.g. convert a TE stmt to C++ and call a system c++ compiler on to see that it compiles - what do you think? (No need to do that in this PR)

Based on off-line discussion, we are going to compile mobilenetv3 into C++, then compile the generated code.

**Comment by cheng-chang** [comment]
Previously, some commit history from master shows up in this PR, which automatically triggers review request from code owners of other components. I reopened the PR at https://github.com/pytorch/pytorch/pull/52980.

**Pull request closed without merging**
